flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapUnionFunction.java

```diff
@@ -0,0 +1,161 @@
+
+import java.lang.invoke.MethodHandle;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.flink.table.api.Expressions.$;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#MAP_UNION}. */
+@Internal
+public class MapUnionFunction extends BuiltInScalarFunction {
+    private final ArrayData.ElementGetter keyElementGetter;
+    private final ArrayData.ElementGetter valueElementGetter;
+
+    private final SpecializedFunction.ExpressionEvaluator keyEqualityEvaluator;
+    private transient MethodHandle keyEqualityHandle;
+
+    public MapUnionFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.MAP_UNION, context);
+        KeyValueDataType outputType =
+                ((KeyValueDataType) context.getCallContext().getOutputDataType().get());
+        final DataType keyDataType = outputType.getKeyDataType();
+        final DataType valueDataType = outputType.getValueDataType();
+        keyElementGetter =
+                ArrayData.createElementGetter(outputType.getKeyDataType().getLogicalType());
+        valueElementGetter =
+                ArrayData.createElementGetter(outputType.getValueDataType().getLogicalType());
+        keyEqualityEvaluator =
+                context.createEvaluator(
+                        $("element1").isEqual($("element2")),
+                        DataTypes.BOOLEAN(),
+                        DataTypes.FIELD("element1", keyDataType.notNull().toInternal()),
+                        DataTypes.FIELD("element2", keyDataType.notNull().toInternal()));
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        keyEqualityHandle = keyEqualityEvaluator.open(context);
+    }
+
+    public @Nullable MapData eval(@Nullable MapData... maps) {
+        try {
+            if (maps == null || maps.length == 0) {
+                return null;
+            }
+            if (maps.length == 1) {
+                return maps[0];
+            }
+            MapData result = maps[0];
+            for (int i = 1; i < maps.length; ++i) {
+                MapData map = maps[i];
+                if (map != null && map.size() > 0) {
+                    result = new MapDataForMapUnion(result, map);
+                }
+            }
+            return result;
+        } catch (Throwable t) {
+            throw new FlinkRuntimeException(t);
+        }
+    }
+
+    private class MapDataForMapUnion implements MapData {
+        private final GenericArrayData keysArray;
+        private final GenericArrayData valuesArray;
+
+        public MapDataForMapUnion(MapData map1, MapData map2) throws Throwable {
+            List<Object> keysList = new ArrayList<>();
+            List<Object> valuesList = new ArrayList<>();
+            boolean isKeyNullExist = false;
+            for (int i = 0; i < map2.size(); i++) {
+                Object key = keyElementGetter.getElementOrNull(map2.keyArray(), i);
```

Comment:
nit: depending on implementation it could generate new objects
so probably `map2.keyArray()` could be extracted into a var before the loop

Reply:
Ok, I will extracted it.